starwars/translate/asm6809-2.12/man/asm6809.1
'\" t
.
.\" ASCII for Unicode ellipsis is three dots
.schar \[u2026] ...
.\" New escape [...] maps to Unicode ellipsis
.char \[...] \[u2026]
.
.\" an-ext.tmac: Check whether we are using grohtml.
.nr mH 0
.if \n(.g \
.  if '\*(.T'html' \
.    nr mH 1
.
.\" What about gropdf?
.nr mP 0
.if \n(.g \
.  if '\*(.T'pdf' \
.    nr mP 1
.
.\" an-ext.tmac: Start example.
.de EX
.  nr mE \\n(.f
.  nf
.  nh
.  ft CW
..
.
.\" an-ext.tmac: End example.
.de EE
.  ft \\n(mE
.  fi
.  hy \\n(HY
..
.
.\" Top level heading; wraps .SH
.de H1
.  if \\n(mP .pdfhref O 1 \\$*
.  SH \\$*
..
.
.\" 2nd level heading; wraps .SS
.de H2
.  if \\n(mP .pdfhref O 2 \\$*
.  SS \\$*
..
.
.\" 3rd level heading; bold font, no indent
.de H3
.  if \\n(.$ \{\
.    if \\n(mP .pdfhref O 3 \\$*
.    .B \&"\\$*"
.  \}
.  br
..
.
.\" Render URL
.de UU
.  ie \\n(mH \{\
\\$1\c
.    do HTML-NS "<a href='\\$2'>"
\\$2\c
.    do HTML-NS "</a>"
\\$3
.  \}
.  el \{\
.    ie \\n(mP \{\
.      pdfhref -W -P "\\$1" -A "\\$3" "\\$2"
.    \}
.    el \{\
\\$1\\$2\\$3
.    \}
.  \}
..
.
.nr PDFOUTLINE.FOLDLEVEL 3
.\"
.pdfview /PageMode /UseOutlines
.pdfinfo /Title asm6809 2.12
.pdfinfo /Author Ciaran Anscomb
.
.TH "asm6809" "1" "February 2019" "asm6809-2.12"
.hy 0
.nh
.H1 NAME
.PP
asm6809\[em]6809 cross-assembler
.H1 SYNOPSIS
.PP
\fBasm6809\fR \[lB]\fIOPTION\fR\[rB]\[...] \[lB]\fISOURCE-FILE\fR\[rB]\[...]
.H1 DESCRIPTION
.PP
\fBasm6809\fR is a portable macro cross assembler targeting the Motorola 6809 and Hitachi 6309 processors. These processors are most commonly encountered in the Dragon and Tandy Colour Computer.
.H1 OPTIONS
.TP
\f(CB\-B\fR, \f(CB\-\-bin\fR
output raw binary file (default)
.TP
\f(CB\-D\fR, \f(CB\-\-dragondos\fR
output DragonDOS binary file
.TP
\f(CB\-C\fR, \f(CB\-\-coco\fR
output CoCo RS-DOS (\[lq]DECB\[rq]) segmented binary file
.TP
\f(CB\-S\fR, \f(CB\-\-srec\fR
output Motorola SREC file
.TP
\f(CB\-H\fR, \f(CB\-\-hex\fR
output Intel hex record file
.TP
\f(CB\-e\fR, \f(CB\-\-exec\fR \fIaddr\fR
EXEC address (for output formats that support one)
.TP
\f(CB\-8\fR, \f(CB\-9\fR, \f(CB\-\-6809\fR
use 6809 ISA (default)
.TP
\f(CB\-3\fR, \f(CB\-\-6309\fR
use 6309 ISA (6809 with extensions)
.TP
\f(CB\-d\fR, \f(CB\-\-define\fR \fIsym\fR\[lB]=\fInumber\fR\[rB]
define a symbol
.TP
\f(CB\-\-setdp\fR \fIvalue\fR
initial value assumed for DP \[lB]undefined\[rB]
.TP
\f(CB\-P\fR, \f(CB\-\-max\-passes\fR \fIn\fR
maximum number of passes to allow symbol values to stabilise \[lB]12\[rB]
.TP
\f(CB\-o\fR, \f(CB\-\-output\fR \fIfile\fR
output filename
.TP
\f(CB\-l\fR, \f(CB\-\-listing\fR \fIfile\fR
create listing file
.TP
\f(CB\-E\fR, \f(CB\-\-exports\fR \fIfile\fR
create exports table
.TP
\f(CB\-s\fR, \f(CB\-\-symbols\fR \fIfile\fR
create symbol table
.TP
\f(CB\-q\fR, \f(CB\-\-quiet\fR
don\[aq]t warn about illegal (but working) code
.TP
\f(CB\-v\fR, \f(CB\-\-verbose\fR
warn about explicitly inefficient code
.TP
\f(CB\-\-help\fR
show help
.TP
\f(CB\-\-version\fR
show program version
.PP
If more than one \fISOURCE-FILE\fR is specified, they are assembled as though they were all in one file.
.H1 USAGE
.PP
Text is read in and parsed, then as many passes are made over the parsed source as necessary (up to a limit), until symbols are resolved and addresses are stable. The fastest or smallest representation should always be chosen where there is ambiguity.
.PP
Output formats are: Raw binary, DragonDOS binary, CoCo RS-DOS (\[lq]DECB\[rq]) binary, Motorola SREC, Intel HEX.
.PP
Additional optional output files are:
.IP \(bu 2
A listing file is an annotated copy of the source file with addresses and generated code prepended to each line.
.IP \(bu 2
An exports file contains a list of all macro definitions and symbols flagged for export with the \f(CBEXPORT\fR pseudo-op. Suitable for inclusion in subsequent source files.
.IP \(bu 2
A symbols file contains a list of \fIall\fR non-local symbols. Suitable for inclusion in subsequent source files, but beware multiple definitions errors if two source files include a common set of symbols.
.PP
Home page: 
.UU "<" "http://www.6809.org.uk/asm6809/" ">"
.H2 Differences to other assemblers
.PP
Motorola syntax allows a comment to follow any operands, separated from them only by whitespace. To an extent, this assembler accepts that, but be aware that as spaces are allowed within expressions, if the comment looks like it is continuing an expression it will generate bad code (or raise an error if the result is syntactically incorrect). Example:
.IP
.EX
0000\ \ 8605\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ lda\ \ \ \ \ #5
0002\ \ C60A\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ ldb\ \ \ \ \ #5\ *\ 2\ \ twice\ first\ number
.EE
.PP
A strict Motorola assembler would generate bytes C6 05 for the second line, as the \[lq]*\~2\[rq] would be ignored. For consistency, it is best to introduce end of line comments with a \f(CB;\fR character. An asterisk (\f(CB*\fR) can introduce whole line comments.
.PP
An unquoted semicolon always introduces a comment. The alternate form of the 6309 instructions \f(CBAIM\fR, \f(CBOIM\fR, etc. listed in some documentation that uses a semicolon as a separator is not accepted.
.PP
A symbol may be forward referenced; any time a reference is unresolvable, another pass is triggered, up to some defined maximum.
.PP
In 6809 indexed addressing, the offset size will default to the fastest possible form, e.g. if the offset is an expression that happens to evaluate to zero, the \fIno offset\fR form will be used. Prepend \f(CB<<\fR to coerce a 5 bit offset, \f(CB<\fR to coerce 8 bits or \f(CB>\fR to coerce 16 bits.
.PP
\fBasm6809\fR currently has no support for OS-9 modules or multiple object linking.
.H2 Program syntax
.PP
Program files are considered line by line. Each line contains up to three fields, separated by whitespace: label, instruction and arguments. An unquoted semicolon (\f(CB;\fR) indicates that the rest of the line is to be considered a comment. Whole line comments may be introduced with an asterisk (\f(CB*\fR). Motorola-style end of line comments without a \f(CB;\fR are accepted, but see the notes about assembler differences.
.PP
Any label must appear at the very beginning of the line. If a label is omitted, whitespace must appear before the operator field. Certain pseudo-ops may affect a label\[aq]s meaning, but usually labels define a symbol referring to the current position in the code (Program Counter, or PC).
.PP
The instruction field contains either an instruction op-code (mnemonic), a pseudo-op (assembler directive), or a macro name for expansion.
.PP
Pseudo-ops allow conditional assembly and inline data, can affect code placement and symbol values and be used to include further files inline. See the section on Pseudo-ops for more information.
.PP
Arguments are a comma-separated list: either instruction operands or arguments to a pseudo-op or macro. Permitted arguments are specific to the instruction or pseudo-op, but in general they may be:
.IP \(bu 2
An expression.
.IP \(bu 2
A register name, with optional pre-decrement or post-increment.
.IP \(bu 2
A nested list surrounded by \f(CB\[lB]\fR and \f(CB\[rB]\fR. This is generally only used to indicate indirect indexed addressing.
.PP
In addition, any argument may be preceded by:
.IP \(bu 2
\f(CB#\fR, indicate immediate value.
.IP \(bu 2
\f(CB<<\fR, force 5-bit index offset.
.IP \(bu 2
\f(CB<\fR, force direct addressing, 8-bit value or 8-bit index offset.
.IP \(bu 2
\f(CB>\fR, force extended addressing, 16-bit value or 16-bit index offset.
.H2 Expressions
.PP
Expressions are formed of:
.IP \(bu 2
A decimal number.
.IP \(bu 2
An octal number preceded by \f(CB@\fR or with a leading \f(CB0\fR.
.IP \(bu 2
A binary number preceded by \f(CB%\fR or \f(CB0b\fR.
.IP \(bu 2
A hexadecimal number preceded by \f(CB$\fR or \f(CB0x\fR.
.IP \(bu 2
A floating point number: decimal digits surrounding exactly one full stop (\f(CB.\fR).
.IP \(bu 2
A single quote followed by any ASCII character (yielding the ASCII value of that character).
.IP \(bu 2
A symbol name, local forward reference or local back reference.
.IP \(bu 2
Any of the above prefixed with a unary minus (\f(CB\-\fR) or plus (\f(CB+\fR).
.IP \(bu 2
A string delimited either by double quotes or \f(CB/\fR.
.IP \(bu 2
A combination of any of the above with arithmetic, bitwise, logical or relational operators.
.IP \(bu 2
Parenthesis to specify precedence.
.PP
The assembler uses multiple passes to resolve expressions. If an expression refers to a symbol that cannot currently be resolved, an extra pass is triggered. Similarly, if a symbol is assigned a value (e.g. by an \f(CBEQU\fR pseudo-op) that differs to its value on the previous pass, another is triggered until it becomes stable.
.PP
When not directly used for their contents (e.g. by \f(CBFCC\fR), strings can be used in place of integer values. The ASCII value of each character is used to represent 8 bits of the integer result up to 32 bits. Example:
.IP
.EX
0000\ \ CC443A\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ ldd\ \ \ \ \ #"D:"
.EE
.H2 Operators
.PP
The following operators are available, listed in descending order of precedence (where operators share a precedence, left-to-right evaluation is performed):
.RS
.TS
tab(;);
c | l.
 \fBOperator\fR ;\fBDescription\fR
.T&
_ | _
c | l.
 \f(CB+\fR ;unary plus
 \f(CB\-\fR ;unary minus
 \f(CB!\fR \f(CB\[ti]\fR ;logical, bitwise NOT
.T&
_ | _
c | l.
 \f(CB*\fR ;multiplication
 \f(CB/\fR ;division
 \f(CB%\fR ;modulo
.T&
_ | _
c | l.
 \f(CB+\fR ;addition
 \f(CB\-\fR ;subtraction
.T&
_ | _
c | l.
 \f(CB<<\fR ;bitwise shift left
 \f(CB>>\fR ;bitwise shift right
.T&
_ | _
c | l.
 \f(CB<\fR \f(CB<=\fR ;relational operators
 \f(CB>\fR \f(CB>=\fR ;relational operators
.T&
_ | _
c | l.
 \f(CB==\fR ;relational equal
 \f(CB!=\fR ;relational not equal
.T&
_ | _
c | l.
 \f(CB&\fR ;bitwise AND
.T&
_ | _
c | l.
 \f(CB\[ha]\fR ;bitwise XOR
.T&
_ | _
c | l.
 \f(CB|\fR ;bitwise OR
.T&
_ | _
c | l.
 \f(CB&&\fR ;logical AND
.T&
_ | _
c | l.
 \f(CB||\fR ;logical OR
.T&
_ | _
c | l.
 \f(CB?:\fR ;ternary operator
.TE
.RE
.PP
Division always returns a floating point result. Other arithmetic operators return integers if both operands are integers, otherwise floating point. Bitwise operators and modulo all cast their operands to integers and return an integer. Relational and logical operators result in 0 if false, 1 if true. Integer calculations are performed using the platform\[aq]s \fIint64_t\fR type, floating point uses \fIdouble\fR.
.H2 Conditional assembly
.PP
The pseudo-ops \f(CBIF\fR, \f(CBELSIF\fR, \f(CBELSE\fR and \f(CBENDIF\fR guide conditional assembly. \f(CBIF\fR and \f(CBELSIF\fR take one argument, which is evaluated as an integer. If the result is non-zero, the following code will be assembled, else it will be skipped. Undefined symbols encountered while evaluating the condition are interpreted as zero (false) rather than raising an error.
.PP
Conditional assembly pseudo-ops are permitted within macro definitions and will be evaluated at the time of expansion, therefore positional variables can be used to affect macro expansion.
.H2 Sections
.PP
Code can be placed into named sections with the \f(CBSECTION\fR pseudo-op. This can make breaking source into multiple input files more comfortable. Without \f(CBORG\fR or \f(CBPUT\fR directives, sections will follow each other in memory in the order they are first defined.
.PP
Within each section, there may exist multiple spans of discontiguous data. Certain output formats are able to represent this, for the others (e.g. DragonDOS), the spans are combined first, with the gaps between them padded with zero bytes.
.H2 Local labels
.PP
Local labels are considered local to the current \fIsection\fR. A local label is any decimal number used in the label field, and the same local label may appear mulitple times, unlike other labels.
.PP
As an operand, a decimal number followed by \f(CBB\fR or \f(CBF\fR is considered to be a back or forward reference to the previous or next occurrence of that numerical local label in the section.
.PP
In this example, the \f(CB1\fR label occurs twice, but each use of \f(CB1B\fR refers to the closest one searching backwards:
.IP
.EX
0000\ \ 8E0400\ \ \ \ scroll\ \ \ \ \ \ ldx\ \ \ \ \ #$0400
0003\ \ EC8820\ \ \ \ 1\ \ \ \ \ \ \ \ \ \ \ ldd\ \ \ \ \ 32,x
0006\ \ ED81\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ std\ \ \ \ \ ,x++
0008\ \ 8C05E0\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ cmpx\ \ \ \ #$05e0
000B\ \ 25F6\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ blo\ \ \ \ \ 1B
000D\ \ CC6060\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ ldd\ \ \ \ \ #$6060
0010\ \ ED81\ \ \ \ \ \ 1\ \ \ \ \ \ \ \ \ \ \ std\ \ \ \ \ ,x++
0012\ \ 8C0600\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ cmpx\ \ \ \ #$0600
0015\ \ 25F9\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ blo\ \ \ \ \ 1B
0017\ \ 39\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ rts
.EE
.PP
An exclamation mark (\f(CB!\fR) in the label field is treated as a local label numbered zero. Operands of \f(CB<\fR and \f(CB>\fR are considered equivalent to \f(CB0B\fR and \f(CB0F\fR respectively, and can therefore refer to the \f(CB!\fR local label. This is included for compatibility with other assemblers.
.PP
As local labels can be repeated, their position is used to distinguish them. For this reason, all file inclusions and macro expansion must occur during the first pass so that the absolute line count at which each local label is encountered remains the same between passes.
.H2 Macros
.PP
Start a macro definition by specifying a name for it in the label field, and \f(CBMACRO\fR in the instruction field. Finish the definition with \f(CBENDM\fR in the instruction field.
.PP
Use a macro by specifying its name in the instruction field. Any arguments given will be available during expansion as a positional variable.
.PP
Positional variables can be used within strings, or pasted to form symbol names. In either case, they must be quoted or they will be passed by value, which will result in an error if they do not correspond to valid symbols by themselves.
.PP

pitrex/changes.txt
Changelog:

7th of May 2021

    added GS BASIC (https://vectrex32.com/)
    BASIC starts in the directory “/basic” – since no directory changes can be done from BASIC, this is the BASIC “home”
    gsbasic.ini for some “option”, more information in the file:
    GS_BASIC.txtm OtherTitle.txt VXTRON.txt
    GS BASIC Copyright (c) 2021 Robert E. Alexander
    Usage and distribution is granted under the MIT licence, see MIT_licence.txt
    THX! Bob!
    also included (amongst others) is VxTRON
    VxTron32 (c) 2020, Jaymz Julian
    in s slightly modified version
    added support of USB keyboards (see source code distribution: usb_readme.txt)
    I started the USB keyboard thing, because I wanted to control “Elite a new Kind” via keyboard, rather than invent tricky button double/tripple usages.
    Elite a new Kind is as of now only included as a “demo” version (start screen), since GS BASIC came along an stole my time 🙂 – actually the whole game is compiled, but I haven’t configured the keyboard in, and didn’t look at/fix any of the non vector gfx yet
    I started porting the MAME z80 processor emulation to be used in AAE, it compiles now without flaws, but I have as yet not changed the API to fit AAE. Once I have done, almost all SEGA games should work.
    I am also thinking about (as a quick hack) to add USB keyboard to AAE, than the default keyboard
    settings should work for all games out of the box, and they would be playable via keyboard – not optimal, but better than “waiting” to manually configuring each and every game to the vectrex controls
    Asteroids (sbt) can now in the ini file be configured to use/not use samples
    during the above changes… some bugs were discovered and fixed in the vectrexInterface, can’t remember them all
    osWrapper.c is “wrappier” than before 🙂 e.g. if keyboard is configured you can use fgetc() to “wait” for keyboard input
    vectrexInterface can be configured to use two sample channels, these will be mixed. Default this is “off” – but it DOES work!
    started looking at bitmaps again, one of the next things will be to add bitmap output to the IRQMode (as of now, this os not yet supported)
    changed the “structure” of the baremetal directories, vectrexInterface (and others) are now build as libraries, not several object files. All Makefiles were changed (slightly)
    I like it, when include files are kept WITH their respective sources, so I did not create a dedicated include directory, the "INC_DIR" must be set in the makefiles to the correct locations
    GS BASIC is written in c++, thus c++ is now supported to be built.
    As an example look at GS BASIC. For now, the global/static constructors must be called “manually” in the main() function (6 lines of code), if you intend to program c++, these lines must be copied!
    added emulation of the TMS5220 to sample out (Star Wars), although since star wars is to slow, the output is rather choppy

12th of April 2021

    added the possibilty to play samples alongside vectordisplay
    (digitized sound will be barely audiable on noBuzz vectrex – there is nothing I can do about it!)
    added for asteroids 2 samples (explosions)
    tweaked battlezone, for “english” settings output might be smoother than before (other languages not supported yet)
    added “1 joystick mode” for battlezone, this is the new default
    for battlezone added dipswitches
    for battlezone highscore saving/loading added
    added sound to battlezone (except the engine sound)
    added the start of AAE emulation the aae emulator can be access from the menu (last arcade emulator)
    then you can chose which game to “play”
    loadable and with “input”: Star Wars (6809) and Major Havoc (6502), but I have not really tried optimizing them – and one can see that!
    Cinematronics games (ccpu) are loadable as well, but I have not configured input
    I forgott to switch IRQ mode on for the AAE emulator, so if configured right, there would be a probably more than 20% speed increase
    I am still waiting on feedback in relation to the Z80 CPU. Once I get a working copy, probably all SEGA games can be supported as well.
    AAE roms are not included – and I don’t know what version you need, but “clues” can be found in the readme in the ROM directory
    many tweaks to the vectrexInterface…

26th of March 2021

    default calibration is back to zero, on the todo list -> calibration for noBuzz vectrex
    added configuration (ini file) for input settings, as of now asteroids only, look at the ini file
    added dipswitch settings (ini file), as of now asteroids only, look at ini file, in game these can be called with button 1+4 on joypad 1
    fixed asteroids output to accomodate the 4 different languages
    fixed a “calibration 0” bug with in the vectrex library, y pos could be always “0” after calibration (visible with asteroids shots)
    fixed “speedy vectrex” raster output
    added ini setting “IN_GAME_SETTINGS” which enables or disables Button 2 of Joypad 2 to invoke settings menu. Per default this is OFF
    vectrex save state, if enabled (ini setting HANDLE_STATES), it asks you whether to save or not, also shows a “SAVING…”.
    Saving seems to take sometimes up to 18 seconds (with my SD cart, less than a second…)

I have to further evaluate the “speed”.
I have measured the piZero speed, and as far as I can measure:
1 Mhz timer is pretty exact
Cycle timer is pretty exact
Vectrex T2 timer is pretty exact
The reason I mention this… although “programmatically” and reports during game support this… the games are running all at the correct speed.
However …

    if I compare the games with MAME or the VecFever – it seems they are to slow
    Kevin mentioned ZBlast as being to slow
    the sample tests… seem output of the samples are to slow

I am at a loss what may cause that. As said timing measurements are correct – somehow I must be doing something wrong here… 

starwars/translate/asm6809-2.12/src/asm6809.c
/*

asm6809, a Motorola 6809 cross assembler
Copyright 2013-2017 Ciaran Anscomb

This program is free software: you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation, either version 3 of the License, or (at your
option) any later version.

*/

/* for getopt_long */
#define _GNU_SOURCE

#include "config.h"

#include <errno.h>
#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "xalloc.h"

#include "asm6809.h"
#include "assemble.h"
#include "error.h"
#include "listing.h"
#include "node.h"
#include "opcode.h"
#include "output.h"
#include "program.h"
#include "section.h"
#include "slist.h"
#include "symbol.h"

struct asm6809_options asm6809_options;

#define OUTPUT_BINARY (0)
#define OUTPUT_DRAGONDOS (1)
#define OUTPUT_COCO (2)
#define OUTPUT_MOTOROLA_SREC (3)
#define OUTPUT_INTEL_HEX (4)

static int max_passes = 12;
static int output_format = OUTPUT_BINARY;
static char *exec_option = NULL;
static char *output_filename = NULL;
static char *exports_filename = NULL;
static char *symbol_filename = NULL;
static char *listing_filename = NULL;
static int isa = asm6809_isa_6809;
static int max_program_depth = 8;
static int setdp = -1;
static int verbosity = 0;

static struct option long_options[] = {
	{ "bin", no_argument, &output_format, OUTPUT_BINARY },
	{ "dragondos", no_argument, &output_format, OUTPUT_DRAGONDOS },
	{ "coco", no_argument, &output_format, OUTPUT_COCO },
	{ "srec", no_argument, &output_format, OUTPUT_MOTOROLA_SREC },
	{ "hex", no_argument, &output_format, OUTPUT_INTEL_HEX },
	{ "exec", required_argument, NULL, 'e' },
	{ "6809", no_argument, &isa, asm6809_isa_6809 },
	{ "6309", no_argument, &isa, asm6809_isa_6309 },
	{ "define", required_argument, NULL, 'd' },
	{ "setdp", required_argument, &setdp, 0 },
	{ "max-passes", required_argument, NULL, 'P' },
	{ "output", required_argument, NULL, 'o' },
	{ "listing", required_argument, NULL, 'l' },
	{ "exports", required_argument, NULL, 'E' },
	{ "symbols", required_argument, NULL, 's' },
	{ "quiet", no_argument, NULL, 'q' },
	{ "verbose", no_argument, NULL, 'v' },
	{ "help", no_argument, NULL, 'h' },
	{ "version", no_argument, NULL, 'V' },
	{ NULL, 0, NULL, 0 }
};

static struct slist *files = NULL;

static struct node *simple_parse_int(const char *);
static void define_symbol(const char *);
static void helptext(void);
static void versiontext(void);
static _Noreturn void tidy_up_and_exit(int status);

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

int main(int argc, char **argv) {

	int c;
	while ((c = getopt_long(argc, argv, "BDCSHe:893d:P:o:l:E:s:qv",
				long_options, NULL)) != -1) {
		switch (c) {
		case 0:
			break;
		case 'B':
			output_format = OUTPUT_BINARY;
			break;
		case 'D':
			output_format = OUTPUT_DRAGONDOS;
			break;
		case 'C':
			output_format = OUTPUT_COCO;
			break;
		case 'S':
			output_format = OUTPUT_MOTOROLA_SREC;
			break;
		case 'H':
			output_format = OUTPUT_INTEL_HEX;
			break;
		case 'e':
			exec_option = optarg;
			break;
		case '8': case '9':
			isa = asm6809_isa_6809;
			break;
		case '3':
			isa = asm6809_isa_6309;
			break;
		case 'd':
			define_symbol(optarg);
			break;
		case 'P':
			{
				long v = strtol(optarg, NULL, 0);
				if (errno != 0 || v < 2 || v > 255) {
					error(error_type_fatal, "invalid value for max-passes");
					error_print_list();
					tidy_up_and_exit(EXIT_FAILURE);
				}
				max_passes = v;
			}
			break;
		case 'o':
			output_filename = optarg;
			break;
		case 'l':
			listing_filename = optarg;
			break;
		case 'E':
			exports_filename = optarg;
			break;
		case 's':
			symbol_filename = optarg;
			break;
		case 'q':
			verbosity = -1;
			break;
		case 'v':
			verbosity = 1;
			break;
		case 'h':
			helptext();
			tidy_up_and_exit(EXIT_SUCCESS);
		case 'V':
			versiontext();
			tidy_up_and_exit(EXIT_SUCCESS);
		default:
			tidy_up_and_exit(EXIT_FAILURE);
		}
	}

	if (optind >= argc) {
		error(error_type_fatal, "no input files");
		error_print_list();
		tidy_up_and_exit(EXIT_FAILURE);
	}

	asm6809_options.isa = isa;
	asm6809_options.max_program_depth = max_program_depth;
	asm6809_options.setdp = setdp;
	asm6809_options.verbosity = verbosity;
	asm6809_options.listing_required = listing_filename ? 1 : 0;

	opcode_init();
	assemble_init();

	/* Read in each file */
	for (int i = optind; i < argc; i++) {
		struct prog *f = prog_new_file(argv[i]);
		files = slist_append(files, f);
	}

	/* Fatal errors? */
	if (error_level >= error_type_syntax) {
		error_print_list();
		tidy_up_and_exit(EXIT_FAILURE);
	}

	/* Attempt to assemble files until consistent */
	for (unsigned pass = 0; pass < max_passes; pass++) {
		error_clear_all();
		listing_free_all();
		section_set("CODE", pass);
		for (struct slist *l = files; l; l = l->next) {
			struct prog *f = l->data;
			assemble_prog(f, pass);
		}
		section_finish_pass();
		/* Only inconsistencies trigger another pass */
		if (error_level != error_type_inconsistent)
			break;
	}

	/* Fatal errors? */
	if (error_level >= error_type_inconsistent) {
		error_print_list();
		tidy_up_and_exit(EXIT_FAILURE);
	}
	/* Otherwise print any warnings */
	error_print_list();

	/* Generate listing file */
	if (listing_filename) {
		FILE *listf = fopen(listing_filename, "wb");
		if (listf) {
			listing_print(listf);
			fclose(listf);
		} else {
			error(error_type_fatal, "%s: %s", listing_filename, strerror(errno));
		}
	}

	/* Special parsing of option exec address option.  Overrides any use of
	 * the END pseudo-op. */
	if (exec_option) {
		struct node *n = simple_parse_int(exec_option);
		if (!n) {
			unsigned v = 0;
			struct node *tmp = symbol_get(exec_option);
			if (tmp) {
				v = tmp->data.as_int & 0xffff;
				node_free(tmp);
			} else {
				error(error_type_fatal, "exec symbol '%s' not defined", exec_option);
			}
			n = node_new_int(v);
		}
		symbol_force_set(".exec", n, 0, max_passes);
	}

	// XXX At the moment listing generation must precede output, as
	// coelescing spans might screw with the span data to which the listing
	// refers.  Not a big deal, but needs fixing.

	/* Generate output file */
	if (output_filename) {
		switch (output_format) {
		case OUTPUT_BINARY:
			output_binary(output_filename);
			break;
		case OUTPUT_DRAGONDOS:
			output_dragondos(output_filename);
			break;
		case OUTPUT_COCO:
			output_coco(output_filename);
			break;
		case OUTPUT_MOTOROLA_SREC:
			output_motorola_srec(output_filename);
			break;
		case OUTPUT_INTEL_HEX:
			output_intel_hex(output_filename);
			break;
		default:
			error(error_type_fatal, "internal: unexpected output format");
			break;
		}
	}

	/* Generate exports file */
	if (exports_filename) {
		FILE *expf = fopen(exports_filename, "wb");
		if (expf) {
			prog_print_exports(expf);
			fclose(expf);
		} else {
			error(error_type_fatal, "%s: %s", exports_filename, strerror(errno));
		}
	}

	/* Generate symbols file */
	if (symbol_filename) {
		FILE *symf = fopen(symbol_filename, "wb");
		if (symf) {
			prog_print_symbols(symf);
			fclose(symf);
		} else {
			error(error_type_fatal, "%s: %s", symbol_filename, strerror(errno));
		}
	}

	/* Any errors in all that? */
	if (error_level >= error_type_syntax) {
		error_print_list();
		tidy_up_and_exit(EXIT_FAILURE);
	}

	error_print_list();
	tidy_up_and_exit(EXIT_SUCCESS);
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

/* Special parsing of arguments.  Integers only for now. */
static struct node *simple_parse_int(const char *str) {
	struct node *n = NULL;
	if (str[0] == '$') {
		n = node_new_int(strtol(str+1, NULL, 16));
	} else if (str[0] == '@') {
		n = node_new_int(strtol(str+1, NULL, 8));
	} else if (str[0] == '%') {
		n = node_new_int(strtol(str+1, NULL, 2));
	} else if (str[0] == '0') {
		if (str[1] == 'x')
			n = node_new_int(strtol(str+2, NULL, 16));
		else if (str[1] == 'b')
			n = node_new_int(strtol(str+2, NULL, 2));
		else
			n = node_new_int(strtol(str, NULL, 10));
	} else if (str[0] >= '0' && str[0] <= '9') {
		n = node_new_int(strtol(str, NULL, 10));
	}
	return n;
}

static void define_symbol(const char *str) {
	char *key = xstrdup(str);
	char *tmp = strchr(key, '=');
	struct node *value;
	if (tmp) {
		*(tmp++) = 0;
		value = simple_parse_int(tmp);
	} else {
		value = node_new_int(1);
	}
	// TODO: check that key is a valid symbol name
	symbol_set(key, value, 0, 0);
	free(key);
	node_free(value);
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

static void helptext(void) {
	puts(
"Usage: asm6809 [OPTION]... SOURCE-FILE...\n"
"Assembles 6809/6309 source code.\n"
"\n"
"  -B, --bin         output to binary file (default)\n"
"  -D, --dragondos   output to DragonDOS binary file\n"
"  -C, --coco        output to CoCo segmented binary file\n"
"  -S, --srec        output to Motorola SREC file\n"
"  -H, --hex         output to Intel hex record file\n"
"  -e, --exec=ADDR   EXEC address (for output formats that support one)\n"
"\n"
"  -8,\n"
"  -9, --6809                  use 6809 ISA (default)\n"
"  -3, --6309                  use 6309 ISA (6809 with extensions)\n"
"  -d, --define=SYM[=NUMBER]   define a symbol\n"
"      --setdp=VALUE           initial value assumed for DP [undefined]\n"
"\n"
"  -o, --output=FILE    set output filename\n"
"  -l, --listing=FILE   create listing file\n"
"  -E, --exports=FILE   create exports table\n"
"  -s, --symbols=FILE   create symbol table\n"
"\n"
"  -q, --quiet     don't warn about illegal (but working) code\n"
"  -v, --verbose   warn about explicitly inefficient code\n"
"\n"
"      --help      show this help\n"
"      --version   show program version\n"
"\n"
"If more than one SOURCE-FILE is specified, they are assembled as though\n"
"they were all in one file."
	    );
}

static void versiontext(void) {
	puts(
"asm6809 " PACKAGE_VERSION "\n"
"Copyright (C) 2018 Ciaran Anscomb\n"
"License: GNU GPL version 3 or later <http://www.gnu.org/licenses/gpl-3.0.html>.\n"
"This is free software: you are free to change and redistribute it.\n"
"There is NO WARRANTY, to the extent permitted by law."
	);
}

/* Call the various free_all routines to tidy up memory, then exit.  Allows us
 * to see what's been missed with valgrind. */

static _Noreturn void tidy_up_and_exit(int status) {
	if (files) {
		slist_free(files);
		files = NULL;
	}
	listing_free_all();
	prog_free_all();
	symbol_free_all();
	section_free_all();
	opcode_free_all();
	assemble_free_all();
	exit(status);
}
